src/constraints.jl

```diff
@@ -540,16 +491,78 @@ function JuMP.build_constraint(
     _error("Invalid set `$set` for logical constraint.")
 end
 
+# Check that logical expression is valid
+function _check_logical_expression(ex)
+    return
+end
+function _check_logical_expression(ex::_LogicalExpr)
+    if !(ex.head in _LogicalOperatorHeads)
+        error("Unrecognized logical operator `$(ex.head)`.")
+    end
+    for a in ex.args
+        _check_logical_expression(a)
+    end
+    return
+end
+
 """
     function JuMP.add_constraint(
-        model::Model,
-        c::ScalarConstraint{<:F, S},
+        model::JuMP.Model,
+        c::JuMP.ScalarConstraint{_LogicalExpr, MOI.EqualTo{Bool}},
         name::String = ""
-    ) where {F <: Union{LogicalVariableRef, _LogicalExpr}, S}
+    )
 
 Extend `JuMP.add_constraint` to allow creating logical proposition constraints 
-for a [`GDPModel`](@ref) with the `@constraint` macro.
+for a [`GDPModel`](@ref) with the `@constraint` macro. Users should define 
+logical constraints via the syntax `@constraint(model, logical_expr := true)`.
+"""
+function JuMP.add_constraint(
+    model::JuMP.Model,
+    c::JuMP.ScalarConstraint{_LogicalExpr, S},
```

Comment:
JuMP doesn't actually pass `EqualTo{Bool}` it passes it wrapped in `_DoNotConvertSet` such that it is not converted to `EqualTo{Float64}`. We cannot explictly type it since `_DoNotConvertSet` is private to JuMP. However, in the function calling `moi_set` returns the underlying `EqualTo{Bool}` set. This is why I have an assertion to make sure the set is correct. Users can only give `EqualTo{Bool}` directly to `add_constraint`, but we encourage them to use `:=` syntax instead to avoid these nuances.

Reply:
So the docstring just shows the user what they can actually use.